gcs_jupyter_plugin/services/gcs.py

```diff
@@ -441,6 +455,93 @@ async def rename_file(self, bucket_name, blob_name, new_name):
             return {"error": str(e), "status": 500}
 
+    async def rename_non_empty_folder(
+        self, bucket: storage.Bucket, source_prefix: str, new_prefix: str
+    ):
+        source_prefix_normalized = (
+            source_prefix if source_prefix.endswith("/") else source_prefix + "/"
+        )
+        new_prefix_normalized = (
+            new_prefix if new_prefix.endswith("/") else new_prefix + "/"
+        )
+
+        blobs_to_rename = list(bucket.list_blobs(prefix=source_prefix_normalized))
+
+        if not blobs_to_rename:
+            empty_folder_blob = bucket.blob(source_prefix_normalized)
+            if empty_folder_blob.exists() and empty_folder_blob.size == 0:
+                self.log.info(
+                    f"Renaming empty folder marker '{source_prefix_normalized}' to '{new_prefix_normalized}'."
+                )
+                new_blob = bucket.rename_blob(empty_folder_blob, new_prefix_normalized)
+                return {
+                    "name": new_blob.name,
+                    "bucket": bucket.name,
+                    "success": True,
+                    "status": 200,
+                    "message": "Empty folder marker renamed.",
+                }
+            else:
+                return {
+                    "error": f"Folder '{source_prefix}' not found or has no objects to rename.",
+                    "status": 404,
+                }
+
+        if (
+            new_prefix_normalized.startswith(source_prefix_normalized)
+            and new_prefix_normalized != source_prefix_normalized
+        ):
+            return {
+                "error": f"Cannot rename folder '{source_prefix}' to a sub-path of itself '{new_prefix}'.",
+                "status": 400,
+            }
+
+        renamed_blobs = []
+        try:
+            for blob in blobs_to_rename:
+                relative_path = blob.name[len(source_prefix_normalized) :]
+                new_blob_name = new_prefix_normalized + relative_path
+
+                if bucket.blob(new_blob_name).exists():
+                    self.log.error(
+                        f"Conflict detected during manual folder rename: '{new_blob_name}' already exists."
+                    )
+                    raise ValueError(
+                        f"Destination object '{new_blob_name}' already exists. Aborting rename."
+                    )
+
+                new_blob = bucket.copy_blob(blob, bucket, new_name=new_blob_name)
+                renamed_blobs.append(new_blob)
+
+            for blob in blobs_to_rename:
+                blob.delete()
+
+            return {
+                "message": f"Folder '{source_prefix}' and its contents renamed to '{new_prefix}'.",
+                "bucket": bucket.name,
+                "success": True,
+                "status": 200,
+            }
+        except Exception as e:
+            self.log.error(
+                f"Error during manual folder rename. Attempting to revert changes: {e}"
```

Comment:
Change "Attempting to revert changes" to "Attempting to restore original files" to make it more clear the limited scope of the rollback.